lte/gateway/c/oai/lib/pipelined_client/PipelinedServiceClient.cpp

```diff
@@ -0,0 +1,713 @@
+/**
+ * Copyright 2020 The Magma Authors.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "PipelinedServiceClient.h"
+
+#include <utility>
+#include <cassert>
+#include <grpcpp/impl/codegen/client_context.h>
+#include <grpcpp/impl/codegen/status.h>
+#include <netinet/in.h>
+#include <cstring>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <thread>
+
+#include <grpcpp/impl/codegen/async_unary_call.h>
+
+#include "lte/protos/pipelined.grpc.pb.h"
+#include "lte/protos/pipelined.pb.h"
+#include "lte/protos/mobilityd.pb.h"
+#include "orc8r/protos/common.pb.h"
+#include "lte/protos/subscriberdb.pb.h"
+#include "ServiceRegistrySingleton.h"
+#include "common_defs.h"
+#include "proto_converters.h"
+
+namespace grpc {
+class Channel;
+class ClientContext;
+class Status;
+}  // namespace grpc
+
+using grpc::Channel;
+using grpc::ChannelCredentials;
+using grpc::ClientContext;
+using grpc::InsecureChannelCredentials;
+using grpc::Status;
+
+namespace magma {
+namespace lte {
+
+PipelinedServiceClient& PipelinedServiceClient::get_instance() {
+  static PipelinedServiceClient client_instance;
+  return client_instance;
+}
+
+PipelinedServiceClient::PipelinedServiceClient() {
+  auto channel = ServiceRegistrySingleton::Instance()->GetGrpcChannel(
+      "pipelined", ServiceRegistrySingleton::LOCAL);
+  stub_ = Pipelined::NewStub(channel);
+  std::thread resp_loop_thread([&]() { rpc_response_loop(); });
+  resp_loop_thread.detach();
+}
+
+//------------------- TUNNEL ADD -------------------
+
+//                    ADD : v4
+//--------------------------------------------------
+int PipelinedServiceClient::UpdateUEIPv4SessionSet(
+    const struct in_addr& ue_ipv4_addr, int vlan, struct in_addr& enb_ipv4_addr,
+    uint32_t in_teid, uint32_t out_teid, const std::string& imsi,
+    uint32_t flow_precedence, const std::string& apn, uint32_t ue_state,
+    std::function<void(Status, UESessionContextResponse)> callback) {
+  UESessionSet request = UESessionSet();
```

Comment:
Please move all these conversion code into proto_convereters.cpp

this function should look like
```
UESessionSet request = <new function in proto_converters.cpp>
UESessionContextResponse response;
PipelinedServiceClient& client = get_instance();

auto local_response = new AsyncLocalResponse<UESessionContextResponse>(
  std::move(callback), RESPONSE_TIMEOUT);

auto response_reader = client.stub_->AsyncUpdateUEState(
  local_response->get_context(), request, &client.queue_);
local_response->set_response_reader(std::move(response_reader));
return RETURNok;
```

Reply:
Comments addressed.